homeassistant/components/motion_blinds/__init__.py

```diff
@@ -106,40 +107,46 @@ async def _async_update_data(self):
     multicast_interface = entry.data.get(CONF_INTERFACE, DEFAULT_INTERFACE)
     wait_for_push = entry.options.get(CONF_WAIT_FOR_PUSH, DEFAULT_WAIT_FOR_PUSH)
 
     entry.async_on_unload(entry.add_update_listener(update_listener))
 
-    # check multicast interface
-    check_multicast_class = ConnectMotionGateway(hass, interface=multicast_interface)
-    working_interface = await check_multicast_class.async_check_interface(host, key)
-    if working_interface != multicast_interface:
-        data = {**entry.data, CONF_INTERFACE: working_interface}
-        hass.config_entries.async_update_entry(entry, data=data)
-        _LOGGER.debug(
-            "Motion Blinds interface updated from %s to %s, "
-            "this should only occur after a network change",
-            multicast_interface,
-            working_interface,
-        )
-
     # Create multicast Listener
-    if KEY_MULTICAST_LISTENER not in hass.data[DOMAIN]:
-        multicast = AsyncMotionMulticast(interface=working_interface)
-        hass.data[DOMAIN][KEY_MULTICAST_LISTENER] = multicast
-        # start listening for local pushes (only once)
-        await multicast.Start_listen()
-
-        # register stop callback to shutdown listening for local pushes
-        def stop_motion_multicast(event):
-            """Stop multicast thread."""
-            _LOGGER.debug("Shutting down Motion Listener")
-            multicast.Stop_listen()
-
-        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_motion_multicast)
+    async with setup_lock:
+        if KEY_MULTICAST_LISTENER not in hass.data[DOMAIN]:
+            # check multicast interface
+            check_multicast_class = ConnectMotionGateway(
+                hass, interface=multicast_interface
+            )
+            working_interface = await check_multicast_class.async_check_interface(
+                host, key
+            )
+            if working_interface != multicast_interface:
+                data = {**entry.data, CONF_INTERFACE: working_interface}
+                hass.config_entries.async_update_entry(entry, data=data)
+                _LOGGER.debug(
+                    "Motion Blinds interface updated from %s to %s, "
+                    "this should only occur after a network change",
+                    multicast_interface,
+                    working_interface,
+                )
+
+            multicast = AsyncMotionMulticast(interface=working_interface)
+            hass.data[DOMAIN][KEY_MULTICAST_LISTENER] = multicast
+            # start listening for local pushes (only once)
+            await multicast.Start_listen()
+
+            # register stop callback to shutdown listening for local pushes
+            def stop_motion_multicast(event):
+                """Stop multicast thread."""
+                _LOGGER.debug("Shutting down Motion Listener")
+                multicast.Stop_listen()
+
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_motion_multicast)
```

Comment:
Probably unsubscribe the listener on config entry unload if it's the last loaded config entry getting unloaded.

Reply:
I think that is already done in `async_unload_entry` right?

Reply:
No we don't unsubscribe the listener for home assistant stop event.

Reply:
Sorry, it took me a while to understand it, but I think I fixed it now in the latest comment.
Thanks for pointing this out!